Bottom: 88240424fbc750c30551387937f2dd7b0c9d79f1
Top:    2202ce9d929ce8b462d3f1c7f7c82e923fd2bd41
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2009-11-11 16:37:14 +0000

Work in progress

---

diff --git a/src/cairo/cairo-bentley-ottmann.c b/src/cairo/cairo-bentley-ottmann.c
index e3b3a0c..03795fb 100644
--- a/src/cairo/cairo-bentley-ottmann.c
+++ b/src/cairo/cairo-bentley-ottmann.c
@@ -46,7 +46,9 @@
 #include <glib.h>
 #include <setjmp.h>
 
+#include "rtree.h"
 #include "polygon.h"
+#include "polygon-priv.h"
 
 #define _cairo_error(x) (x)
 
@@ -73,6 +75,8 @@ struct _cairo_bo_edge {
     cairo_bo_edge_t *prev;
     cairo_bo_edge_t *next;
     cairo_point_t middle;
+    PLINE *p;
+    VNODE *v;
 };
 
 /* the parent is always given by index/2 */
@@ -1054,8 +1058,11 @@ _cairo_bo_event_queue_insert_stop (cairo_bo_event_queue_t        *event_queue,
     cairo_bo_point32_t point;
 
     point.y = edge->edge.bottom;
-    point.x = _line_compute_intersection_x_for_y (&edge->edge.line,
-                                                  point.y);
+    if (edge->edge.line.p1.y == edge->edge.line.p2.y)
+        point.x = edge->edge.line.p2.x;
+    else
+        point.x = _line_compute_intersection_x_for_y (&edge->edge.line,
+                                                      point.y);
     return _cairo_bo_event_queue_insert (event_queue,
                                          CAIRO_BO_EVENT_TYPE_STOP,
                                          edge, NULL,
@@ -1342,14 +1349,21 @@ _add_result_edge (cairo_array_t *array,
        edge->top = tmp;
     }
 
+#if 0
     printf ("Emitting result edge (%i,%i)-(%i,%i)\n",
             edge->line.p1.x, edge->line.p1.y,
             edge->line.p2.x, edge->line.p2.y);
+#else
+//    printf ("\tLine[%i %i %i %i 1500 2000 \"clearline\"]\n",
+//            edge->line.p1.x, edge->line.p1.y,
+//            edge->line.p2.x, edge->line.p2.y);
+#endif
 
 //    return _cairo_array_append (array, edge);
     return CAIRO_STATUS_SUCCESS;
 }
 
+static void do_intersect (cairo_bo_edge_t *e1, cairo_bo_edge_t *e2, cairo_point_t point);
 
 /* Execute a single pass of the Bentley-Ottmann algorithm on edges,
  * generating trapezoids according to the fill_rule and appending them
@@ -1471,14 +1485,17 @@ _cairo_bentley_ottmann_tessellate_bo_edges (cairo_bo_event_t   **start_events,
         case CAIRO_BO_EVENT_TYPE_INTERSECTION:
             e1 = ((cairo_bo_queue_event_t *) event)->e1;
             e2 = ((cairo_bo_queue_event_t *) event)->e2;
-            _cairo_bo_event_queue_delete (&event_queue, event);
 
             /* skip this intersection if its edges are not adjacent */
-            if (e2 != e1->next)
+            if (e2 != e1->next) {
+                printf ("Breaking because edges not adjacent - will we return?\n");
                 break;
+            }
 
             intersection_count++;
 
+            do_intersect (e1, e2, event->point);
+
             {
                 cairo_edge_t intersected;
                 /* FIXME: Coordinates of the intersection?? */
@@ -1502,6 +1519,8 @@ _cairo_bentley_ottmann_tessellate_bo_edges (cairo_bo_event_t   **start_events,
                 e2->middle = event->point;
             }
 
+            _cairo_bo_event_queue_delete (&event_queue, event);
+
             left = e1->prev;
             right = e2->next;
 
@@ -2051,10 +2070,13 @@ my_cairo_test (void)
   GList *data = NULL;
   int i = 0;
 
+  return;
+
   printf ("Cairo bentley ottmann test\n");
 
   points = g_new0 (bos_line, 4);
 
+#if 0
   /* Line from (10,10)-(20,20) */
   points[i].a.x = 10; points[i].a.y = 10;
   points[i].b.x = 20; points[i].b.y = 20;
@@ -2075,6 +2097,21 @@ my_cairo_test (void)
   points[i].num = i;
   data = g_list_prepend (data, &points[i]);
   i++;
+#endif
+
+  /* Line from (10,10)-(20,20) */
+  points[i].a.x = 10; points[i].a.y = 10;
+  points[i].b.x = 20; points[i].b.y = 20;
+  points[i].num = i;
+  data = g_list_prepend (data, &points[i]);
+  i++;
+
+  /* Line from (15,20)-(15,10) */
+  points[i].a.x = 15; points[i].a.y = 10;
+  points[i].b.x = 15; points[i].b.y = 20;
+  points[i].num = i;
+  data = g_list_prepend (data, &points[i]);
+  i++;
 
 #if 0
   /* Line from (14,10)-(16,20) */
@@ -2095,7 +2132,62 @@ my_cairo_test (void)
   bentley_ottmann_intersect_segments (data);
 }
 
-#ifdef OLD_WAY
+static int
+vect_equal (cairo_point_t v1, Vector v2)
+{
+  return (v1.x == v2[0] && v1.y == v2[1]);
+}				/* vect_equal */
+
+static inline void
+cntrbox_adjust (PLINE * c, cairo_point_t p)
+{
+  c->xmin = min (c->xmin, p.x);
+  c->xmax = max (c->xmax, p.x + 1);
+  c->ymin = min (c->ymin, p.y);
+  c->ymax = max (c->ymax, p.y + 1);
+}
+
+/*
+ * adjust_tree()
+ * (C) 2006 harry eaton
+ * This replaces the segment in the tree with the two new segments after
+ * a vertex has been added
+ */
+static int
+adjust_tree (PLINE *p, VNODE *v)
+{
+  struct seg *s = lookup_seg (p, v);
+  struct seg *q;
+
+  q = malloc (sizeof (struct seg));
+  if (!q)
+    return 1;
+  if (s->v != v)
+    printf ("FUBAR1\n");
+  q->v = s->v;
+  if (s->p != p)
+    printf ("FUBAR2\n");
+  q->p = s->p;
+  q->box.X1 = min (q->v->point[0], q->v->next->point[0]);
+  q->box.X2 = max (q->v->point[0], q->v->next->point[0]) + 1;
+  q->box.Y1 = min (q->v->point[1], q->v->next->point[1]);
+  q->box.Y2 = max (q->v->point[1], q->v->next->point[1]) + 1;
+  r_insert_entry (p->tree, (const BoxType *) q, 1);
+  q = malloc (sizeof (struct seg));
+  if (!q)
+    return 1;
+  q->v = s->v->next;
+  q->p = s->p;
+  q->box.X1 = min (q->v->point[0], q->v->next->point[0]);
+  q->box.X2 = max (q->v->point[0], q->v->next->point[0]) + 1;
+  q->box.Y1 = min (q->v->point[1], q->v->next->point[1]);
+  q->box.Y2 = max (q->v->point[1], q->v->next->point[1]) + 1;
+  r_insert_entry (p->tree, (const BoxType *) q, 1);
+  r_delete_entry (p->tree, (const BoxType *) s);
+  free (s);
+  return 0;
+}
+
 
 /*
 node_add
@@ -2109,9 +2201,10 @@ node_add
  4 means the intersection was not on the dest point
 */
 static VNODE *
-node_add (VNODE * dest, Vector po, int *new_point)
+node_add (VNODE * dest, cairo_point_t po, int *new_point)
 {
   VNODE *p;
+  Vector v;
 
   if (vect_equal (po, dest->point))
     return dest;
@@ -2120,7 +2213,8 @@ node_add (VNODE * dest, Vector po, int *new_point)
       (*new_point) += 4;
       return dest->next;
     }
-  p = poly_CreateNode (po);
+  v[0] = po.x;  v[1] = po.y;
+  p = poly_CreateNode (v);
   if (p == NULL)
     return NULL;
   (*new_point) += 5;
@@ -2131,6 +2225,27 @@ node_add (VNODE * dest, Vector po, int *new_point)
   return (dest->next = dest->next->prev = p);
 }				/* node_add */
 
+static VNODE *
+node_add_single (VNODE * dest, cairo_point_t po)
+{
+  VNODE *p;
+  Vector v;
+
+  if (vect_equal (po, dest->point))
+    return dest;
+  if (vect_equal (po, dest->next->point))
+    return dest->next;
+  v[0] = po.x;  v[1] = po.y;
+  p = poly_CreateNode (v);
+  if (p == NULL)
+    return NULL;
+  p->prev = dest;
+  p->next = dest->next;
+  p->cvc_prev = p->cvc_next = NULL;
+  p->Flags.status = UNKNWN;
+  return (dest->next = dest->next->prev = p);
+}				/* node_add */
+
 
 /*
 node_add_point
@@ -2141,7 +2256,7 @@ node_add_point
 */
 
 static int
-node_add_point (VNODE * a, VNODE * b, Vector p)
+node_add_point (VNODE * a, VNODE * b, cairo_point_t p)
 {
   int res = 0;
 
@@ -2158,48 +2273,251 @@ node_add_point (VNODE * a, VNODE * b, Vector p)
   return res;
 }				/* node_add_point */
 
+static VNODE *
+node_add_single_point (VNODE *inp_node, cairo_point_t p)
+{
+  VNODE *out_node;
 
+  /* JUST A HUNCH: */
+//  inp_node->cvc_prev = inp_node->cvc_next = (CVCList *) - 1;
+
+  out_node = node_add_single (inp_node, p);
+  out_node->cvc_prev = out_node->cvc_next = (CVCList *) - 1;
+
+  if (out_node == inp_node ||
+      out_node == inp_node->next) {
+    /* No node was added - apparently it already existed */
+    return NULL;
+  }
+
+  return out_node;
+}				/* node_add_point */
 
 
 static void
-dummy (void)
+do_intersect (cairo_bo_edge_t *e1, cairo_bo_edge_t *e2, cairo_point_t point)
 {
-  struct info *i = (struct info *) cl;
-  struct seg *s = (struct seg *) b;
-  Vector s1, s2;
-  int cnt, res;
-
-  i->s->p->Flags.status = ISECTED;
-  s->p->Flags.status = ISECTED;
-  for (; cnt; cnt--)
-    {
-      res = node_add_point (i->v, s->v, cnt > 1 ? s2 : s1);
-      if (res < 0)
-        return 1;                /* error */
-      /* adjust the bounding box and tree if necessary */
-      if (res & 2) {
-        cntrbox_adjust (i->s->p, cnt > 1 ? s2 : s1);
-        if (adjust_tree (i->s->p->tree, i->s))
-          return 1;
-      }
-      /* if we added a node in the tree we need to change the tree */
-      if (res & 1) {
-        cntrbox_adjust (s->p, cnt > 1 ? s2 : s1);
-        if (adjust_tree (i->tree, s))
-          return 1;
-      }
-      if (res & 3) {               /* if a point was inserted start over */
-        DEBUGP ("new intersection at (%d, %d)\n", cnt > 1 ? s2[0] : s1[0], cnt > 1 ? s2[1] : s1[1]);
-        longjmp (*i->env, 1);
-      }
+  VNODE *new_node;
+
+  // i->s is some seg from the edge tree, pointing to v (and corresponds to one of the intersected items)
+  // s is some other seg the item hit.
+  // Need to know the VNODE of the segment (VNODE)-----(VNODE->next) and the PLINE they belong to.
+
+//  cnt = vect_inters2 (e2->v->point, e2->v->next->point,
+//                      e1->v->point, e1->v->next->point, s1, s2);
+
+  // cnt == 0: No intersection (error)
+  // cnt == 1: Intersection, s1 gives the intersection point
+  // cnt == 2: (GUESS) Lines coincident: -----X======x----
+  //                                          ^_s1   ^_s2
+
+  // BUT: We know our _single_ intersection is passed as (point.x,point.y) - so why bother?
+  // Just need to check that intersection isn't with one of our existing vertex end-points?
+
+  e1->p->Flags.status = ISECTED;
+  e2->p->Flags.status = ISECTED;
+
+//  if (cnt == 0) {
+//    printf ("Alleged no intersection error\n");
+//  }
+//  if (cnt == 2) {
+//    printf ("ALLEGED TWO INTERSECTIONS ERROR");
+//  }
+
+  new_node = node_add_single_point (e1->v, point);
+  /* adjust the bounding box and tree if necessary */
+  if (new_node != NULL) {
+    e1->p->Count ++; /* ??? */
+    cntrbox_adjust (e1->p, point);
+    if (adjust_tree (e1->p, e1->v)) return; /* error */
+    /* Need to decide whether the new piece, or the old piece is
+       going to continue seeing "action" in the sweepline algorithm */
+#if 0
+    if (new_node->point[1] > e1->v->point[1]) {
+      e1->v = new_node;
+    } else if (new_node->point[1] == e1->v->point[1]) {
+      if (new_node->point[0] > e1->v->point[0])
+        e1->v = new_node;
+    }
+#endif
+//    e1->v = new_node;
+  }
+  /* if we added a node in the tree we need to change the tree */
+  new_node = node_add_single_point (e2->v, point);
+  if (new_node != NULL) {
+    e2->p->Count ++; /* ??? */
+    cntrbox_adjust (e2->p, point);
+    if (adjust_tree (e2->p, e2->v)) return /*1*/;
+#if 0
+    if (new_node->point[1] > e2->v->point[1]) {
+      e2->v = new_node;
+    } else if (new_node->point[1] == e2->v->point[1]) {
+      if (new_node->point[0] > e2->v->point[0])
+        e2->v = new_node;
     }
-  return 0;
-}
 #endif
+//    e2->v = new_node;
+  }
+  return;
+}
+
+
+static void
+poly_area_to_start_events (POLYAREA                *poly,
+                           cairo_bo_start_event_t  *events,
+                           cairo_bo_event_t       **event_ptrs,
+                           int                     *counter)
+{
+    int i = *counter;
+    PLINE *contour;
+
+    /* Loop over contours */
+    for (contour = poly->contours; contour != NULL; contour = contour->next) {
+      /* Loop over nodes, adding edges */
+      VNODE *bv;
+      bv = &contour->head;
+      do {
+        int x1, y1, x2, y2;
+        cairo_edge_t cairo_edge;
+        /* Node is between bv->point[0,1] and bv->next->point[0,1] */
+
+        /* HACK TEST: */
+        bv->cvc_prev = bv->cvc_next = NULL;
+//        bv->cvc_prev = bv->cvc_next = (CVCList *) - 1;
+
+        if (bv->point[1] == bv->next->point[1]) {
+            if (bv->point[0] < bv->next->point[0]) {
+              x1 = bv->point[0];
+              y1 = bv->point[1];
+              x2 = bv->next->point[0];
+              y2 = bv->next->point[1];
+            } else {
+              x1 = bv->next->point[0];
+              y1 = bv->next->point[1];
+              x2 = bv->point[0];
+              y2 = bv->point[1];
+            }
+        } else if (bv->point[1] < bv->next->point[1]) {
+          x1 = bv->point[0];
+          y1 = bv->point[1];
+          x2 = bv->next->point[0];
+          y2 = bv->next->point[1];
+        } else {
+          x1 = bv->next->point[0];
+          y1 = bv->next->point[1];
+          x2 = bv->point[0];
+          y2 = bv->point[1];
+        }
+
+        cairo_edge.line.p1.x = x1;
+        cairo_edge.line.p1.y = cairo_edge.top = y1;
+        cairo_edge.line.p2.x = x2;
+        cairo_edge.line.p2.y = cairo_edge.bottom = y2;
+        cairo_edge.dir = 0;
+
+        event_ptrs[i] = (cairo_bo_event_t *) &events[i];
+
+        events[i].type = CAIRO_BO_EVENT_TYPE_START;
+        events[i].point.y = cairo_edge.top;
+        events[i].point.x =
+            _line_compute_intersection_x_for_y (&cairo_edge.line,
+                                                events[i].point.y);
+
+        events[i].edge.edge = cairo_edge;
+        events[i].edge.prev = NULL;
+        events[i].edge.next = NULL;
+        events[i].edge.p = contour;
+        events[i].edge.v = bv;
+        i++;
+
+      } while ((bv = bv->next) != &contour->head);
+    }
+
+    *counter = i;
+}
 
 
 int
 bo_intersect (jmp_buf *jb, POLYAREA *b, POLYAREA *a)
 {
-  return 0;
+
+    int intersections;
+    cairo_status_t status;
+    cairo_bo_start_event_t stack_events[CAIRO_STACK_ARRAY_LENGTH (cairo_bo_start_event_t)];
+    cairo_bo_start_event_t *events;
+    cairo_bo_event_t *stack_event_ptrs[ARRAY_LENGTH (stack_events) + 1];
+    cairo_bo_event_t **event_ptrs;
+    int num_events;
+    int i, j, k;
+    VNODE *doh;
+    cairo_traps_t *traps = NULL;
+
+    PLINE *contour;
+
+    num_events = 0;
+
+    j = 0; k = 0;
+    for (contour = a->contours; contour != NULL; contour = contour->next) {
+        int tmp = 0;
+        j += contour->Count;
+        doh = &contour->head;
+        do {tmp++;} while ((doh = doh->next) != &contour->head);
+        k+= tmp;
+        contour->Count = tmp;
+    }
+    num_events += k;
+    if (k != j)
+      printf ("OH CRAPPY DOODLE, j=%i, k=%i\n", j, k);
+
+    j = 0; k = 0;
+    for (contour = b->contours; contour != NULL; contour = contour->next) {
+        int tmp = 0;
+        j += contour->Count;
+        doh = &contour->head;
+        do {tmp++;} while ((doh = doh->next) != &contour->head);
+        k+= tmp;
+        contour->Count = tmp;
+    }
+    if (k != j)
+      printf ("OH CRAPPY DOODLE, j=%i, k=%i\n", j, k);
+    num_events += MAX(j,k);
+
+    if (unlikely (0 == num_events))
+        return CAIRO_STATUS_SUCCESS;
+
+    events = stack_events;
+    event_ptrs = stack_event_ptrs;
+    if (num_events > ARRAY_LENGTH (stack_events)) {
+        events = _cairo_malloc_ab_plus_c (num_events,
+                                          sizeof (cairo_bo_start_event_t) +
+                                          sizeof (cairo_bo_event_t *),
+                                          sizeof (cairo_bo_event_t *));
+        if (unlikely (events == NULL))
+            return _cairo_error (CAIRO_STATUS_NO_MEMORY);
+
+        event_ptrs = (cairo_bo_event_t **) (events + num_events);
+    }
+
+    i = 0;
+
+    poly_area_to_start_events (a, events, event_ptrs, &i);
+    poly_area_to_start_events (b, events, event_ptrs, &i);
+
+    /* XXX: This would be the convenient place to throw in multiple
+     * passes of the Bentley-Ottmann algorithm. It would merely
+     * require storing the results of each pass into a temporary
+     * cairo_traps_t. */
+    status = _cairo_bentley_ottmann_tessellate_bo_edges (event_ptrs,
+                                                         num_events,
+                                                         traps,
+                                                         &intersections);
+    printf ("Number of intersections was %i\n", intersections);
+#if DEBUG_TRAPS
+    dump_traps (traps, "bo-polygon-out.txt");
+#endif
+
+    if (events != stack_events)
+        free (events);
+
+    return status;
 }
diff --git a/src/polygon-priv.h b/src/polygon-priv.h
new file mode 100644
index 0000000..c27606b
--- /dev/null
+++ b/src/polygon-priv.h
@@ -0,0 +1,24 @@
+#define ISECTED 3
+#define UNKNWN  0
+#define INSIDE  1
+#define OUTSIDE 2
+#define SHARED 3
+#define SHARED2 4
+
+#define TOUCHES 99
+
+#define ISECT_BAD_PARAM (-1)
+#define ISECT_NO_MEMORY (-2)
+
+typedef struct seg
+{
+  BoxType box;
+  VNODE *v;
+  PLINE *p;
+  int intersected;
+} seg;
+
+
+struct seg *lookup_seg (PLINE *contour, VNODE *vertex);
+
+int vect_inters2 (Vector p1, Vector p2, Vector q1, Vector q2, Vector S1, Vector S2);
diff --git a/src/polygon1.c b/src/polygon1.c
index d6e9b91..8cc0758 100644
--- a/src/polygon1.c
+++ b/src/polygon1.c
@@ -83,14 +83,8 @@ int vect_inters2 (Vector A, Vector B, Vector C, Vector D, Vector S1,
 /* note that a vertex v's Flags.status represents the edge defined by
  * v to v->next (i.e. the edge is forward of v)
  */
-#define ISECTED 3
-#define UNKNWN  0
-#define INSIDE  1
-#define OUTSIDE 2
-#define SHARED 3
-#define SHARED2 4
 
-#define TOUCHES 99
+#include "polygon-priv.h"
 
 #define NODE_LABEL(n)  ((n)->Flags.status)
 #define LABEL_NODE(n,l) ((n)->Flags.status = (l))
@@ -198,11 +192,6 @@ node_add_single (VNODE * dest, Vector po)
   p->Flags.status = UNKNWN;
   return p;
 }				/* node_add */
-
-#define ISECT_BAD_PARAM (-1)
-#define ISECT_NO_MEMORY (-2)
-
-
 /*
 new_descriptor
   (C) 2006 harry eaton
@@ -492,14 +481,6 @@ cntrbox_adjust (PLINE * c, Vector p)
 
 /* some structures for handling segment intersections using the rtrees */
 
-typedef struct seg
-{
-  BoxType box;
-  VNODE *v;
-  PLINE *p;
-  int intersected;
-} seg;
-
 typedef struct insert_task
 {
   VNODE *new_node;
@@ -610,12 +591,11 @@ seg_in_seg (const BoxType * b, void *cl)
 //  printf ("Looking at intersection between %p(%i) and %p(%i)\n",
 //          s->v, s->v->Flags.intersected, i->v, i->v->Flags.intersected);
 
-  if (s->intersected || i->s->intersected)
-    {
+  if (s->intersected || i->s->intersected) {
 //    printf ("Need to restart intersection\n");
-      i->need_restart = 1;
-      return 0;
-    }
+    i->need_restart = 1;
+    return 0;
+  }
 
   cnt = vect_inters2 (s->v->point, s->v->next->point,
 		      i->v->point, i->v->next->point, s1, s2);
@@ -630,35 +610,34 @@ seg_in_seg (const BoxType * b, void *cl)
       int done_insert = 0;
       new_node = node_add_single_point (i->v, cnt > 1 ? s2 : s1);
       if (new_node != NULL)
-	{
-	  insert_task *task = g_new0 (insert_task, 1);
-	  task->new_node = new_node;
-	  task->seg = i->s;
-	  task->seg->intersected = 1;
-	  i->node_insert_list = g_list_prepend (i->node_insert_list, task);
-	  done_insert = 1;
-	}
+        {
+          insert_task *task = g_new0 (insert_task, 1);
+          task->new_node = new_node;
+          task->seg = i->s;
+          task->seg->intersected = 1;
+          i->node_insert_list = g_list_prepend (i->node_insert_list, task);
+          done_insert = 1;
+        }
 //      else
 //        printf (".");
       new_node = node_add_single_point (s->v, cnt > 1 ? s2 : s1);
       if (new_node != NULL)
-	{
-	  insert_task *task = g_new0 (insert_task, 1);
-	  task->new_node = new_node;
-	  task->seg = s;
-	  task->seg->intersected = 1;
-	  i->node_insert_list = g_list_prepend (i->node_insert_list, task);
-	  return 0;		/* Don't do any more processing */
-	}
+        {
+          insert_task *task = g_new0 (insert_task, 1);
+          task->new_node = new_node;
+          task->seg = s;
+          task->seg->intersected = 1;
+          i->node_insert_list = g_list_prepend (i->node_insert_list, task);
+          return 0; /* Don't do any more processing */
+        }
 //      else
 //        printf (":");
-      if (done_insert)
-	{
+      if (done_insert) {
 //        printf ("Long-jmping back, since we intersected on i\n");
-	  longjmp (*i->env, 1);	/* Skip this contour if we intersected on i */
-	  i->need_restart = 1;	/* If we skip some processing, we definately need a restart */
-	  return 0;
-	}
+        longjmp (*i->env, 1); /* Skip this contour if we intersected on i */
+        i->need_restart = 1; /* If we skip some processing, we definately need a restart */
+        return 0;
+      }
     }
   return 0;
 }
@@ -715,6 +694,25 @@ get_seg (const BoxType * b, void *cl)
   return 0;
 }
 
+struct seg *
+lookup_seg (PLINE *contour, VNODE *vertex)
+{
+  struct info info;
+  BoxType box;
+
+  info.v = vertex;
+  box.X2 = (box.X1 = vertex->point[0]) + 1;
+  box.Y2 = (box.Y1 = vertex->point[1]) + 1;
+  /* fill in the segment in info corresponding to this node */
+  if (setjmp (info.sego) == 0)
+    {
+      r_search_1 (contour->tree, &box, NULL, get_seg, &info);
+      assert (0);
+    }
+  return info.s;
+}
+
+
 /*
  * intersect() (and helpers)
  * (C) 2006, harry eaton
@@ -736,9 +734,9 @@ get_seg (const BoxType * b, void *cl)
 static int
 contour_bounds_touch (const BoxType * b, void *cl)
 {
-  contour_info *c_info = (contour_info *) cl;
+  contour_info *c_info = (contour_info *)cl;
   PLINE *pa = c_info->pa;
-  PLINE *pb = (PLINE *) b;
+  PLINE *pb = (PLINE *)b;
   PLINE *rtree_over;
   PLINE *looping_over;
   VNODE *av;			/* node iterators */
@@ -759,17 +757,17 @@ contour_bounds_touch (const BoxType * b, void *cl)
    */
   if (pa->Count < pb->Count)
     {
-      rtree_over = pb;
+      rtree_over   = pb;
       looping_over = pa;
     }
   else
     {
-      rtree_over = pa;
+      rtree_over   = pa;
       looping_over = pb;
     }
 
   av = &looping_over->head;
-  do				/* Loop over the nodes in the smaller contour */
+  do  /* Loop over the nodes in the smaller contour */
     {
       /* check this edge for any insertions */
       double dx;
@@ -777,56 +775,54 @@ contour_bounds_touch (const BoxType * b, void *cl)
 #if 0
       /* We know we will just reject any intersections found until
          the next pass anyway */
-      if (av->Flags.intersected)
-	{
+      if (av->Flags.intersected) {
 //        printf ("skip\n");
-	  continue;
-	}
+        continue;
+      }
 #endif
 
       info.v = av;
       /* compute the slant for region trimming */
       dx = av->next->point[0] - av->point[0];
       if (dx == 0)
-	info.m = 0;
+        info.m = 0;
       else
-	{
-	  info.m = (av->next->point[1] - av->point[1]) / dx;
-	  info.b = av->point[1] - info.m * av->point[0];
-	}
+        {
+          info.m = (av->next->point[1] - av->point[1]) / dx;
+          info.b = av->point[1] - info.m * av->point[0];
+        }
       box.X2 = (box.X1 = av->point[0]) + 1;
       box.Y2 = (box.Y1 = av->point[1]) + 1;
 
       /* fill in the segment in info corresponding to this node */
       if (setjmp (info.sego) == 0)
-	{
-	  r_search (looping_over->tree, &box, NULL, get_seg, &info);
-	  assert (0);
-	}
+        {
+          r_search (looping_over->tree, &box, NULL, get_seg, &info);
+          assert (0);
+        }
 
       /* If we're going to have another pass anyway, skip this */
       if (info.s->intersected && info.node_insert_list != NULL)
-	continue;
+        continue;
 
       if (setjmp (restart))
-	continue;
+        continue;
 
-      /* NB: If this actually hits anything, we are teleported back to the beginning */
-      info.tree = rtree_over->tree;
-      if (info.tree)
-	if (UNLIKELY (r_search (info.tree, &info.s->box,
-				seg_in_region, seg_in_seg, &info)))
-	  exit (0);
-//            return err_no_memory;     /* error */
+        /* NB: If this actually hits anything, we are teleported back to the beginning */
+        info.tree = rtree_over->tree;
+        if (info.tree)
+          if (UNLIKELY (r_search (info.tree, &info.s->box,
+                                  seg_in_region, seg_in_seg, &info)))
+            exit (0);
+//            return err_no_memory;	/* error */
     }
   while ((av = av->next) != &looping_over->head);
 
   c_info->node_insert_list = info.node_insert_list;
-  if (info.need_restart)
-    {
+  if (info.need_restart) {
 //    printf ("info.needs_restart says we need to restart\n");
-      c_info->need_restart = 1;
-    }
+    c_info->need_restart = 1;
+  }
   return 0;
 }
 
@@ -842,7 +838,7 @@ insert_new_nodes_cb (gpointer data, gpointer userdata)
   task->new_node->next = task->seg->v->next;
   task->seg->v->next->prev = task->new_node;
   task->seg->v->next = task->new_node;
-  task->seg->p->Count++;
+  task->seg->p->Count ++;
 
   cntrbox_adjust (task->seg->p, task->new_node->point);
   if (adjust_tree (task->seg->p->tree, task->seg))
@@ -870,9 +866,9 @@ intersect_impl (jmp_buf * jb, POLYAREA * b, POLYAREA * a, int add)
       a = t;
     }
 
-//  setjmp (c_info.restart);            /* we loop back here whenever a vertex is inserted */
+//  setjmp (c_info.restart);		/* we loop back here whenever a vertex is inserted */
 
-  for (pa = a->contours; pa; pa = pa->next)	/* Loop over the contours of POLYAREA "a" */
+  for (pa = a->contours; pa; pa = pa->next)     /* Loop over the contours of POLYAREA "a" */
     {
       BoxType sb;
       jmp_buf out;
@@ -882,16 +878,16 @@ intersect_impl (jmp_buf * jb, POLYAREA * b, POLYAREA * a, int add)
       c_info.pa = pa;
 
       if (!add)
-	{
-	  retval = setjmp (out);
-	  if (retval)
-	    {
-	      /* The intersection test short-circuited back here,
-	       * we need to clean up, then longjmp to jb */
-	      longjmp (*jb, retval);
-	    }
-	  c_info.getout = &out;
-	}
+        {
+          retval = setjmp (out);
+          if (retval)
+            {
+              /* The intersection test short-circuited back here,
+               * we need to clean up, then longjmp to jb */
+              longjmp (*jb, retval);
+            }
+          c_info.getout = &out;
+        }
 
       sb.X1 = pa->xmin;
       sb.Y1 = pa->ymin;
@@ -899,20 +895,18 @@ intersect_impl (jmp_buf * jb, POLYAREA * b, POLYAREA * a, int add)
       sb.Y2 = pa->ymax + 1;
 
       r_search (b->contour_tree, &sb, NULL, contour_bounds_touch, &c_info);
-      if (c_info.need_restart)
-	{
-	  need_restart = 1;
+      if (c_info.need_restart) {
+        need_restart = 1;
 //        printf ("contour_bounds_touch: need_restart\n");
 //        break;
-	}
+      }
     }
 
 //  printf ("----\n");
-  if (c_info.node_insert_list != NULL)
-    {
+  if (c_info.node_insert_list != NULL) {
 //    printf ("Processing %i new nodes\n", g_list_length (c_info.node_insert_list));
-      need_restart = 1;		/* Any new nodes could intersect */
-    }
+    need_restart = 1; /* Any new nodes could intersect */
+  }
   g_list_foreach (c_info.node_insert_list, insert_new_nodes_cb, NULL);
   g_list_free (c_info.node_insert_list);
 //  printf ("====\n");
@@ -924,8 +918,7 @@ static int
 intersect (jmp_buf * jb, POLYAREA * b, POLYAREA * a, int add)
 {
   int call_count = 1;
-  while (intersect_impl (jb, b, a, add))
-    call_count++;
+  while (intersect_impl (jb, b, a, add)) call_count++;
   return 0;
 }
 
@@ -947,16 +940,22 @@ M_POLYAREA_intersect (jmp_buf * e, POLYAREA * afst, POLYAREA * bfst, int add)
 	      a->contours->xmin <= b->contours->xmax &&
 	      a->contours->ymin <= b->contours->ymax)
 	    {
-	      if (!add)
-		{
-		  if (UNLIKELY (intersect (e, a, b, add)))
-		    error (err_no_memory);
-		}
-	      else
-		{
-		  if (UNLIKELY (bo_intersect (e, a, b)))
-		    error (err_no_memory);
-		}
+#if 1
+              if (!add)
+                {
+                  if (UNLIKELY (intersect (e, a, b, add))) {
+                    printf ("is_enom\n");
+                    error (err_no_memory);
+                  }
+                }
+              else
+#endif
+                {
+                  if (UNLIKELY (bo_intersect (e, a, b))) {
+                    printf ("bo_is_enom\n");
+                    error (err_no_memory);
+                  }
+                }
 	    }
 	}
       while (add && (a = a->f) != afst);
@@ -964,8 +963,10 @@ M_POLYAREA_intersect (jmp_buf * e, POLYAREA * afst, POLYAREA * bfst, int add)
 	if (curcB->Flags.status == ISECTED)
 	  {
 	    the_list = add_descriptors (curcB, 'B', the_list);
-	    if (UNLIKELY (the_list == NULL))
+	    if (UNLIKELY (the_list == NULL)) {
+              printf ("add_desc\n");
 	      error (err_no_memory);
+            }
 	  }
     }
   while (add && (b = b->f) != bfst);
@@ -975,8 +976,10 @@ M_POLYAREA_intersect (jmp_buf * e, POLYAREA * afst, POLYAREA * bfst, int add)
 	if (curcA->Flags.status == ISECTED)
 	  {
 	    the_list = add_descriptors (curcA, 'A', the_list);
-	    if (UNLIKELY (the_list == NULL))
+	    if (UNLIKELY (the_list == NULL)) {
+              printf ("add_desc2\n");
 	      error (err_no_memory);
+            }
 	  }
     }
   while (add && (a = a->f) != afst);
@@ -995,7 +998,7 @@ cntrbox_inside (PLINE * c1, PLINE * c2)
 /* Routines for making labels */
 
 static int
-count_contours_i_am_inside (const BoxType * b, void *cl)
+count_contours_i_am_inside (const BoxType *b, void *cl)
 {
   PLINE *me = cl;
   PLINE *check = (PLINE *) b;
@@ -1033,22 +1036,19 @@ cntr_in_M_POLYAREA (PLINE * poly, POLYAREA * outfst, BOOLp test)
 	break;
       outer = (POLYAREA *) heap_remove_smallest (heap);
 
-      switch (r_search
-	      (outer->contour_tree, (BoxType *) poly, NULL,
-	       count_contours_i_am_inside, poly))
-	{
-	case 0:		/* Didn't find anything in this piece, Keep looking */
-	  break;
-	case 1:		/* Found we are inside this piece, and not any of its holes */
-	  heap_destroy (&heap);
-	  return TRUE;
-	case 2:		/* Found inside a hole in the smallest polygon so far. No need to check the other polygons */
-	  heap_destroy (&heap);
-	  return FALSE;
-	default:
-	  printf ("Something strange here\n");
-	  break;
-	}
+      switch (r_search (outer->contour_tree, (BoxType *)poly, NULL, count_contours_i_am_inside, poly)) {
+        case 0: /* Didn't find anything in this piece, Keep looking */
+          break;
+        case 1: /* Found we are inside this piece, and not any of its holes */
+          heap_destroy (&heap);
+          return TRUE;
+        case 2: /* Found inside a hole in the smallest polygon so far. No need to check the other polygons */
+          heap_destroy (&heap);
+          return FALSE;
+        default:
+          printf ("Something strange here\n");
+          break;
+      }
     }
   while (1);
   heap_destroy (&heap);
@@ -1112,28 +1112,28 @@ label_contour (PLINE * a)
   do
     {
       if (cur->cvc_next)	/* _always_ examine cross vertex */
-	{
-	  label = node_label (cur);
-	  if (first_not_unknown == NULL)
-	    first_not_unknown = cur;
-	  continue;
-	}
+        {
+          label = node_label (cur);
+          if (first_not_unknown == NULL)
+            first_not_unknown = cur;
+          continue;
+        }
 
       if (NODE_LABEL (cur) != UNKNWN)
-	{
-	  /* This might get hit if the last cross vertex labelling
-	   * pass was a bit keen, and labeled past its own edges?
-	   */
-	  label = NODE_LABEL (cur);
-	  continue;
-	}
+        {
+          /* This might get hit if the last cross vertex labelling
+           * pass was a bit keen, and labeled past its own edges?
+           */
+          label = NODE_LABEL (cur);
+          continue;
+        }
 
       if (label == INSIDE || label == OUTSIDE)
-	{
-	  /* This labels nodes which aren't cross-connected */
-	  LABEL_NODE (cur, label);
-	  continue;
-	}
+        {
+          /* This labels nodes which aren't cross-connected */
+          LABEL_NODE (cur, label);
+          continue;
+        }
     }
   while ((cur = cur->next) != first_not_unknown);
 #ifdef DEBUG_ALL_LABELS
@@ -1171,11 +1171,10 @@ M_POLYAREA_label_separated (PLINE * afst, POLYAREA * b, BOOLp touch)
 {
   PLINE *curc = afst;
 
-  for (curc = afst; curc != NULL; curc = curc->next)
-    {
-      if (cntr_label_POLYAREA (curc, b, touch) && touch)
-	return TRUE;
-    }
+  for (curc = afst; curc != NULL; curc = curc->next) {
+    if (cntr_label_POLYAREA (curc, b, touch) && touch)
+      return TRUE;
+  }
   return FALSE;
 }
 
@@ -1228,7 +1227,7 @@ InsCntr (jmp_buf * e, PLINE * c, POLYAREA ** dst)
 
 static void
 PutContour (jmp_buf * e, PLINE * cntr, POLYAREA ** contours, PLINE ** holes,
-	    POLYAREA * owner, POLYAREA * parent, PLINE * parent_contour)
+            POLYAREA *owner, POLYAREA * parent, PLINE * parent_contour)
 {
   assert (cntr != NULL);
   assert (cntr->Count > 2);
@@ -1237,7 +1236,7 @@ PutContour (jmp_buf * e, PLINE * cntr, POLYAREA ** contours, PLINE ** holes,
   if (cntr->Flags.orient == PLF_DIR)
     {
       if (owner != NULL)
-	r_delete_entry (owner->contour_tree, (BoxType *) cntr);
+        r_delete_entry (owner->contour_tree, (BoxType *)cntr);
       InsCntr (e, cntr, contours);
     }
   /* put hole into temporary list */
@@ -1248,28 +1247,28 @@ PutContour (jmp_buf * e, PLINE * cntr, POLYAREA ** contours, PLINE ** holes,
 	{
 	  cntr->next = parent_contour->next;
 	  parent_contour->next = cntr;
-	  if (owner != parent)
-	    {
-	      if (owner != NULL)
-		r_delete_entry (owner->contour_tree, (BoxType *) cntr);
-	      r_insert_entry (parent->contour_tree, (BoxType *) cntr, 0);
-	    }
+          if (owner != parent)
+            {
+              if (owner != NULL)
+                r_delete_entry (owner->contour_tree, (BoxType *)cntr);
+              r_insert_entry (parent->contour_tree, (BoxType *)cntr, 0);
+            }
 	}
       else
 	{
 	  cntr->next = *holes;
 	  *holes = cntr;	/* let cntr be 1st hole in list */
-	  /* We don't insert the holes into an r-tree,
-	   * they just form a linked list */
-	  if (owner != NULL)
-	    r_delete_entry (owner->contour_tree, (BoxType *) cntr);
+          /* We don't insert the holes into an r-tree,
+           * they just form a linked list */
+          if (owner != NULL)
+            r_delete_entry (owner->contour_tree, (BoxType *)cntr);
 	}
     }
 }				/* PutContour */
 
 static inline void
-remove_contour (POLYAREA * piece, PLINE * prev_contour, PLINE * contour,
-		int remove_rtree_entry)
+remove_contour (POLYAREA *piece, PLINE *prev_contour, PLINE *contour,
+                int remove_rtree_entry)
 {
   if (piece->contours == contour)
     piece->contours = contour->next;
@@ -1282,7 +1281,7 @@ remove_contour (POLYAREA * piece, PLINE * prev_contour, PLINE * contour,
   contour->next = NULL;
 
   if (remove_rtree_entry)
-    r_delete_entry (piece->contour_tree, (BoxType *) contour);
+    r_delete_entry (piece->contour_tree, (BoxType *)contour);
 }
 
 struct polyarea_info
@@ -1295,37 +1294,34 @@ static int
 heap_it (const BoxType * b, void *cl)
 {
   heap_t *heap = (heap_t *) cl;
-  struct polyarea_info *pa_info = (struct polyarea_info *) b;
+  struct polyarea_info *pa_info = (struct polyarea_info *)b;
   PLINE *p = pa_info->pa->contours;
   if (p->Count == 0)
-    return 0;			/* how did this happen? */
+    return 0;  /* how did this happen? */
   heap_insert (heap, p->area, pa_info);
   return 1;
 }
 
-struct find_inside_info
-{
+struct find_inside_info {
   jmp_buf jb;
   PLINE *want_inside;
   PLINE *result;
 };
 
 static int
-find_inside (const BoxType * b, void *cl)
+find_inside (const BoxType *b, void *cl)
 {
   struct find_inside_info *info = cl;
   PLINE *check = (PLINE *) b;
   /* Do test on check to see if it inside info->want_inside */
   /* If it is: */
-  if (check->Flags.orient == PLF_DIR)
-    {
-      return 0;
-    }
-  if (poly_ContourInContour (info->want_inside, check))
-    {
-      info->result = check;
-      longjmp (info->jb, 1);
-    }
+  if (check->Flags.orient == PLF_DIR) {
+    return 0;
+  }
+  if (poly_ContourInContour (info->want_inside, check)) {
+    info->result = check;
+    longjmp (info->jb, 1);
+  }
   return 0;
 }
 
@@ -1349,7 +1345,7 @@ InsertHoles (jmp_buf * e, POLYAREA * dest, PLINE ** src)
   curc = dest;
   do
     {
-      num_polyareas++;
+      num_polyareas ++;
     }
   while ((curc = curc->f) != dest);
 
@@ -1397,7 +1393,7 @@ InsertHoles (jmp_buf * e, POLYAREA * dest, PLINE ** src)
       pa_info = heap_remove_smallest (heap);
       if (heap_is_empty (heap))
 	{			/* only one possibility it must be the right one */
-	  assert (poly_ContourInContour (pa_info->pa->contours, curh));
+          assert (poly_ContourInContour (pa_info->pa->contours, curh));
 	  container = pa_info->pa->contours;
 	}
       else
@@ -1430,48 +1426,42 @@ InsertHoles (jmp_buf * e, POLYAREA * dest, PLINE ** src)
 	}
       else
 	{
-	  /* Need to check if this new hole means we need to kick out any old ones for reprocessing */
-	  while (1)
-	    {
-	      struct find_inside_info info;
-	      PLINE *prev;
-
-	      info.want_inside = curh;
-
-	      /* Set jump return */
-	      if (setjmp (info.jb))
-		{
-		  /* Returned here! */
-		}
-	      else
-		{
-		  info.result = NULL;
-		  /* Rtree search, calling back a routine to longjmp back with data about any hole inside the added one */
-		  /*   Be sure not to bother jumping back to report the main contour! */
-		  r_search (pa_info->pa->contour_tree, (BoxType *) curh, NULL,
-			    find_inside, &info);
-
-		  /* Nothing found? */
-		  break;
-		}
-
-	      /* We need to find the contour before it, so we can update its next pointer */
-	      printf ("Found inside??\n");
+          /* Need to check if this new hole means we need to kick out any old ones for reprocessing */
+          while (1) {
+            struct find_inside_info info;
+            PLINE *prev;
+
+            info.want_inside = curh;
+
+            /* Set jump return */
+            if (setjmp (info.jb)) {
+              /* Returned here! */
+            } else {
+              info.result = NULL;
+              /* Rtree search, calling back a routine to longjmp back with data about any hole inside the added one */
+              /*   Be sure not to bother jumping back to report the main contour! */
+              r_search (pa_info->pa->contour_tree, (BoxType *)curh, NULL, find_inside, &info);
+
+              /* Nothing found? */
+              break;
+            }
+
+            /* We need to find the contour before it, so we can update its next pointer */
+            printf ("Found inside??\n");
 //            exit (0);
-	      prev = container;
-	      while (prev->next != info.result)
-		{
-		  prev = prev->next;
-		}
+            prev = container;
+            while (prev->next != info.result) {
+              prev = prev->next;
+            }
 
-	      /* Remove hole from the contour */
-	      remove_contour (pa_info->pa, prev, info.result, TRUE);
+            /* Remove hole from the contour */
+            remove_contour (pa_info->pa, prev, info.result, TRUE);
 
-	      /* Add hole as the next on the list to be processed in this very function */
-	      info.result->next = *src;
-	      *src = info.result;
-	    }
-	  /* End check for kicked out holes */
+            /* Add hole as the next on the list to be processed in this very function */
+            info.result->next = *src;
+            *src = info.result;
+          }
+          /* End check for kicked out holes */
 
 	  /* link at front of hole list */
 	  curh->next = container->next;
@@ -1692,37 +1682,37 @@ Gather (VNODE * start, PLINE ** result, J_Rule v_rule, DIRECTION initdir)
 }				/* Gather */
 
 static void
-Collect1 (jmp_buf * e, VNODE * cur, DIRECTION dir, POLYAREA ** contours,
-	  PLINE ** holes, J_Rule j_rule)
+Collect1 (jmp_buf * e, VNODE *cur, DIRECTION dir, POLYAREA **contours, PLINE ** holes, J_Rule j_rule)
 {
   PLINE *p = NULL;		/* start making contour */
   int errc = err_ok;
-  if ((errc =
-       Gather (dir == FORW ? cur : cur->next, &p, j_rule, dir)) != err_ok)
-    {
-      if (p != NULL)
-	poly_DelContour (&p);
-      error (errc);
-    }
-  if (!p)
-    return;
-  poly_PreContour (p, TRUE);
-  if (p->Count > 2)
-    {
+	if ((errc =
+	     Gather (dir == FORW ? cur : cur->next, &p, j_rule,
+		     dir)) != err_ok)
+	  {
+	    if (p != NULL)
+	      poly_DelContour (&p);
+	    error (errc);
+	  }
+	if (!p)
+	  return;
+	poly_PreContour (p, TRUE);
+	if (p->Count > 2)
+	  {
 #ifdef DEBUG_GATHER
-      DEBUGP ("adding contour with %d verticies and direction %c\n",
-	      p->Count, p->Flags.orient ? 'F' : 'B');
+	    DEBUGP ("adding contour with %d verticies and direction %c\n",
+		    p->Count, p->Flags.orient ? 'F' : 'B');
 #endif
-      PutContour (e, p, contours, holes, NULL, NULL, NULL);
-    }
-  else
-    {
-      /* some sort of computation error ? */
+	    PutContour (e, p, contours, holes, NULL, NULL, NULL);
+	  }
+	else
+	  {
+	    /* some sort of computation error ? */
 #ifdef DEBUG_GATHER
-      DEBUGP ("Bad contour! Not enough points!!\n");
+	    DEBUGP ("Bad contour! Not enough points!!\n");
 #endif
-      poly_DelContour (&p);
-    }
+	    poly_DelContour (&p);
+	  }
 }
 
 static void
@@ -1734,21 +1724,20 @@ Collect (jmp_buf * e, PLINE * a, POLYAREA ** contours, PLINE ** holes,
 
   cur = &a->head;
   do
-    {
-      if (s_rule (cur, &dir) && cur->Flags.mark == 0)
-	Collect1 (e, cur, dir, contours, holes, j_rule);
-      other = cur;
-      if ((other->cvc_prev && jump (&other, &dir, j_rule)))
-	Collect1 (e, other, dir, contours, holes, j_rule);
-    }
+   {
+    if (s_rule (cur, &dir) && cur->Flags.mark == 0)
+        Collect1(e, cur, dir, contours, holes, j_rule);
+    other = cur;
+    if ((other->cvc_prev && jump(&other, &dir, j_rule)))
+        Collect1(e, other, dir, contours, holes, j_rule);
+   }
   while ((cur = cur->next) != &a->head);
 }				/* Collect */
 
 
 static int
 cntr_Collect (jmp_buf * e, PLINE ** A, POLYAREA ** contours, PLINE ** holes,
-	      int action, POLYAREA * owner, POLYAREA * parent,
-	      PLINE * parent_contour)
+	      int action, POLYAREA *owner, POLYAREA * parent, PLINE *parent_contour)
 {
   PLINE *tmprev;
 
@@ -1796,7 +1785,7 @@ cntr_Collect (jmp_buf * e, PLINE ** A, POLYAREA ** contours, PLINE ** holes,
 	      PutContour (e, tmprev, contours, holes, owner, NULL, NULL);
 	      return TRUE;
 	    }
-	  /* break; *//* Fall through (I think this is correct! pcjc2) */
+	  /* break; */ /* Fall through (I think this is correct! pcjc2) */
 	case PBO_UNITE:
 	case PBO_SUB:
 	  if ((*A)->Flags.status == OUTSIDE)
@@ -1805,8 +1794,7 @@ cntr_Collect (jmp_buf * e, PLINE ** A, POLYAREA ** contours, PLINE ** holes,
 	      /* disappear this contour (rtree entry removed in PutContour) */
 	      *A = tmprev->next;
 	      tmprev->next = NULL;
-	      PutContour (e, tmprev, contours, holes, owner, parent,
-			  parent_contour);
+	      PutContour (e, tmprev, contours, holes, owner, parent, parent_contour);
 	      return TRUE;
 	    }
 	  break;
@@ -1872,21 +1860,21 @@ M_B_AREA_Collect (jmp_buf * e, POLYAREA * bfst, POLYAREA ** contours,
 
 
 static inline int
-contour_is_first (POLYAREA * a, PLINE * cur)
+contour_is_first (POLYAREA *a, PLINE *cur)
 {
   return (a->contours == cur);
 }
 
 
 static inline int
-contour_is_last (PLINE * cur)
+contour_is_last (PLINE *cur)
 {
   return (cur->next == NULL);
 }
 
 
 static inline void
-remove_polyarea (POLYAREA ** list, POLYAREA * piece)
+remove_polyarea (POLYAREA **list, POLYAREA *piece)
 {
   /* If this item was the start of the list, advance that pointer */
   if (*list == piece)
@@ -1903,7 +1891,7 @@ remove_polyarea (POLYAREA ** list, POLYAREA * piece)
 
 static void
 M_POLYAREA_separate_isected (jmp_buf * e, POLYAREA ** pieces,
-			     PLINE ** holes, PLINE ** isected)
+                             PLINE ** holes, PLINE ** isected)
 {
   POLYAREA *a = *pieces;
   POLYAREA *anext;
@@ -1914,88 +1902,75 @@ M_POLYAREA_separate_isected (jmp_buf * e, POLYAREA ** pieces,
     return;
 
   /* TODO: STASH ENOUGH INFORMATION EARLIER ON, SO WE CAN REMOVE THE INTERSECTED
-     CONTOURS WITHOUT HAVING TO WALK THE FULL DATA-STRUCTURE LOOKING FOR THEM. */
+           CONTOURS WITHOUT HAVING TO WALK THE FULL DATA-STRUCTURE LOOKING FOR THEM. */
 
-  do
-    {
-      int hole_contour = 0;
-      int is_outline = 1;
+  do {
+    int hole_contour = 0;
+    int is_outline = 1;
 
-      anext = a->f;
-      finished = (anext == *pieces);
+    anext = a->f;
+    finished = (anext == *pieces);
 
-      prev = NULL;
-      for (curc = a->contours; curc != NULL; curc = next, is_outline = 0)
-	{
-	  int is_first = contour_is_first (a, curc);
-	  int is_last = contour_is_last (curc);
-	  int isect_contour = (curc->Flags.status == ISECTED);
-
-	  next = curc->next;
+    prev = NULL;
+    for (curc = a->contours; curc != NULL; curc = next, is_outline = 0) {
+      int is_first = contour_is_first (a, curc);
+      int is_last = contour_is_last (curc);
+      int isect_contour = (curc->Flags.status == ISECTED);
 
-	  if (isect_contour || hole_contour)
-	    {
+      next = curc->next;
 
-	      /* Reset the intersection flags, since we keep these pieces */
-	      if (curc->Flags.status != ISECTED)
-		curc->Flags.status = UNKNWN;
+      if (isect_contour || hole_contour) {
 
-	      remove_contour (a, prev, curc, !(is_first && is_last));
+        /* Reset the intersection flags, since we keep these pieces */
+        if (curc->Flags.status != ISECTED)
+          curc->Flags.status = UNKNWN;
 
-	      if (isect_contour)
-		{
-		  /* Link into the list of intersected contours */
-		  curc->next = *isected;
-		  *isected = curc;
-		}
-	      else if (hole_contour)
-		{
-		  /* Link into the list of holes */
-		  curc->next = *holes;
-		  *holes = curc;
-		}
-	      else
-		{
-		  assert (0);
-		}
+        remove_contour (a, prev, curc, !(is_first && is_last));
 
-	      if (is_first && is_last)
-		{
-		  remove_polyarea (pieces, a);
-		  poly_Free (&a);	/* NB: Sets a to NULL */
-		}
+        if (isect_contour) {
+          /* Link into the list of intersected contours */
+          curc->next = *isected;
+          *isected = curc;
+        } else if (hole_contour) {
+          /* Link into the list of holes */
+          curc->next = *holes;
+          *holes = curc;
+        } else {
+          assert (0);
+        }
 
-	    }
-	  else
-	    {
-	      /* Note the item we just didn't delete as the next
-	         candidate for having its "next" pointer adjusted.
-	         Saves walking the contour list when we delete one. */
-	      prev = curc;
-	    }
+        if (is_first && is_last) {
+          remove_polyarea (pieces, a);
+          poly_Free (&a); /* NB: Sets a to NULL */
+        }
 
-	  /* If we move or delete an outer contour, we need to move any holes
-	     we wish to keep within that contour to the holes list. */
-	  if (is_outline && isect_contour)
-	    hole_contour = 1;
+      } else {
+        /* Note the item we just didn't delete as the next
+           candidate for having its "next" pointer adjusted.
+           Saves walking the contour list when we delete one. */
+        prev = curc;
+      }
 
-	}
+      /* If we move or delete an outer contour, we need to move any holes
+         we wish to keep within that contour to the holes list. */
+      if (is_outline && isect_contour)
+        hole_contour = 1;
 
-      /* If we deleted all the pieces of the polyarea, *pieces is NULL */
     }
-  while ((a = anext), *pieces != NULL && !finished);
+
+    /* If we deleted all the pieces of the polyarea, *pieces is NULL */
+  } while ((a = anext), *pieces !=NULL && !finished);
 }
 
 
-struct find_inside_m_pa_info
-{
+struct find_inside_m_pa_info {
   jmp_buf jb;
   POLYAREA *want_inside;
   PLINE *result;
 };
 
 static int
-find_inside_m_pa (const BoxType * b, void *cl)
+find_inside_m_pa (const BoxType *b, void *cl)
 {
   struct find_inside_m_pa_info *info = cl;
   PLINE *check = (PLINE *) b;
@@ -2005,18 +1980,17 @@ find_inside_m_pa (const BoxType * b, void *cl)
   /* Don't look at contours marked as being intersected */
   if (check->Flags.status == ISECTED)
     return 0;
-  if (cntr_in_M_POLYAREA (check, info->want_inside, FALSE))
-    {
-      info->result = check;
-      longjmp (info->jb, 1);
-    }
+  if (cntr_in_M_POLYAREA (check, info->want_inside, FALSE)) {
+    info->result = check;
+    longjmp (info->jb, 1);
+  }
   return 0;
 }
 
 
 static void
 M_POLYAREA_update_primary (jmp_buf * e, POLYAREA ** pieces,
-			   PLINE ** holes, int action, POLYAREA * bpa)
+                           PLINE ** holes, int action, POLYAREA *bpa)
 {
   POLYAREA *a = *pieces;
   POLYAREA *b;
@@ -2031,8 +2005,7 @@ M_POLYAREA_update_primary (jmp_buf * e, POLYAREA ** pieces,
   if (a == NULL)
     return;
 
-  switch (action)
-    {
+  switch (action) {
     case PBO_ISECT:
       del_outside = 1;
       break;
@@ -2040,213 +2013,184 @@ M_POLYAREA_update_primary (jmp_buf * e, POLYAREA ** pieces,
     case PBO_SUB:
       del_inside = 1;
       break;
-    case PBO_XOR:		/* NOT IMPLEMENTED OR USED */
+    case PBO_XOR: /* NOT IMPLEMENTED OR USED */
       inv_inside = 1;
       assert (0);
       break;
-    }
+  }
 
-  box = *((BoxType *) bpa->contours);
+  box = *((BoxType *)bpa->contours);
   b = bpa;
-  while ((b = b->f) != bpa)
-    {
-      BoxType *b_box = (BoxType *) b->contours;
-      MAKEMIN (box.X1, b_box->X1);
-      MAKEMIN (box.Y1, b_box->Y1);
-      MAKEMAX (box.X2, b_box->X2);
-      MAKEMAX (box.Y2, b_box->Y2);
-    }
-
-  if (del_inside)
-    {
-
-      do
-	{
-	  anext = a->f;
-	  finished = (anext == *pieces);
-
-	  /* Test the outer contour first, as we may need to remove all children */
-
-	  /* We've not yet split intersected contours out, just ignore them */
-	  if (a->contours->Flags.status != ISECTED &&
-	      /* Pre-filter on bounding box */
-	      ((a->contours->xmin >= box.X1) && (a->contours->ymin >= box.Y1)
-	       && (a->contours->xmax <= box.X2)
-	       && (a->contours->ymax <= box.Y2)) &&
-	      /* Then test properly */
-	      cntr_in_M_POLYAREA (a->contours, bpa, FALSE))
-	    {
-
-	      /* Delete this contour, all children -> holes queue */
-
-	      /* Delete the outer contour */
-	      curc = a->contours;
-	      remove_contour (a, NULL, curc, FALSE);	/* Rtree deleted in poly_Free below */
-	      /* a->contours now points to the remaining holes */
-	      poly_DelContour (&curc);
-
-	      if (a->contours != NULL)
-		{
-		  /* Find the end of the list of holes */
-		  curc = a->contours;
-		  while (curc->next != NULL)
-		    curc = curc->next;
-
-		  /* Take the holes and prepend to the holes queue */
-		  curc->next = *holes;
-		  *holes = a->contours;
-		  a->contours = NULL;
-		}
-
-	      remove_polyarea (pieces, a);
-	      poly_Free (&a);	/* NB: Sets a to NULL */
-
-	      continue;
-	    }
-
-	  /* Loop whilst we find INSIDE contours to delete */
-	  while (1)
-	    {
-	      struct find_inside_m_pa_info info;
-	      PLINE *prev;
-
-	      info.want_inside = bpa;
-
-	      /* Set jump return */
-	      if (setjmp (info.jb))
-		{
-		  /* Returned here! */
-		}
-	      else
-		{
-		  info.result = NULL;
-		  /* r-tree search, calling back a routine to longjmp back with
-		   * data about any hole inside the B polygon.
-		   * NB: Does not jump back to report the main contour!
-		   */
-		  r_search (a->contour_tree, &box, NULL, find_inside_m_pa,
-			    &info);
-
-		  /* Nothing found? */
-		  break;
-		}
+  while ((b = b->f) != bpa) {
+    BoxType *b_box = (BoxType *)b->contours;
+    MAKEMIN (box.X1, b_box->X1);
+    MAKEMIN (box.Y1, b_box->Y1);
+    MAKEMAX (box.X2, b_box->X2);
+    MAKEMAX (box.Y2, b_box->Y2);
+  }
 
-	      /* We need to find the contour before it, so we can update its next pointer */
-	      prev = a->contours;
-	      while (prev->next != info.result)
-		{
-		  prev = prev->next;
-		}
+  if (del_inside) {
 
-	      /* Remove hole from the contour */
-	      remove_contour (a, prev, info.result, TRUE);
-	      poly_DelContour (&info.result);
-	    }
-	  /* End check for deleted holes */
-
-	  /* If we deleted all the pieces of the polyarea, *pieces is NULL */
-	}
-      while ((a = anext), *pieces != NULL && !finished);
+    do {
+      anext = a->f;
+      finished = (anext == *pieces);
 
-      return;
-    }
-  else
-    {
-      /* This path isn't optimised for speed */
-    }
+      /* Test the outer contour first, as we may need to remove all children */
+
+      /* We've not yet split intersected contours out, just ignore them */
+      if (a->contours->Flags.status != ISECTED &&
+          /* Pre-filter on bounding box */
+          ((a->contours->xmin >= box.X1) && (a->contours->ymin >= box.Y1) &&
+           (a->contours->xmax <= box.X2) && (a->contours->ymax <= box.Y2)) &&
+          /* Then test properly */
+          cntr_in_M_POLYAREA (a->contours, bpa, FALSE)) {
+
+        /* Delete this contour, all children -> holes queue */
+
+        /* Delete the outer contour */
+        curc = a->contours;
+        remove_contour (a, NULL, curc, FALSE); /* Rtree deleted in poly_Free below */
+        /* a->contours now points to the remaining holes */
+        poly_DelContour (&curc);
+
+        if (a->contours != NULL) {
+          /* Find the end of the list of holes */
+          curc = a->contours;
+          while (curc->next != NULL)
+            curc = curc->next;
+
+          /* Take the holes and prepend to the holes queue */
+          curc->next = *holes;
+          *holes = a->contours;
+          a->contours = NULL;
+        }
+
+        remove_polyarea (pieces, a);
+        poly_Free (&a); /* NB: Sets a to NULL */
+
+        continue;
+      }
+
+      /* Loop whilst we find INSIDE contours to delete */
+      while (1) {
+        struct find_inside_m_pa_info info;
+        PLINE *prev;
+
+        info.want_inside = bpa;
+
+        /* Set jump return */
+        if (setjmp (info.jb)) {
+          /* Returned here! */
+        } else {
+          info.result = NULL;
+          /* r-tree search, calling back a routine to longjmp back with
+           * data about any hole inside the B polygon.
+           * NB: Does not jump back to report the main contour!
+           */
+          r_search (a->contour_tree, &box, NULL, find_inside_m_pa, &info);
+
+          /* Nothing found? */
+          break;
+        }
+
+        /* We need to find the contour before it, so we can update its next pointer */
+        prev = a->contours;
+        while (prev->next != info.result) {
+          prev = prev->next;
+        }
+
+        /* Remove hole from the contour */
+        remove_contour (a, prev, info.result, TRUE);
+        poly_DelContour (&info.result);
+      }
+      /* End check for deleted holes */
+
+    /* If we deleted all the pieces of the polyarea, *pieces is NULL */
+    } while ((a = anext), *pieces != NULL && !finished);
 
-  do
-    {
-      int hole_contour = 0;
-      int is_outline = 1;
+    return;
+  } else {
+    /* This path isn't optimised for speed */
+  }
 
-      anext = a->f;
-      finished = (anext == *pieces);
+  do {
+    int hole_contour = 0;
+    int is_outline = 1;
 
-      prev = NULL;
-      for (curc = a->contours; curc != NULL; curc = next, is_outline = 0)
-	{
-	  int is_first = contour_is_first (a, curc);
-	  int is_last = contour_is_last (curc);
-	  int del_contour = 0;
+    anext = a->f;
+    finished = (anext == *pieces);
 
-	  next = curc->next;
+    prev = NULL;
+    for (curc = a->contours; curc != NULL; curc = next, is_outline = 0) {
+      int is_first = contour_is_first (a, curc);
+      int is_last = contour_is_last (curc);
+      int del_contour = 0;
 
-	  if (del_outside)
-	    del_contour = curc->Flags.status != ISECTED &&
-	      !cntr_in_M_POLYAREA (curc, bpa, FALSE);
+      next = curc->next;
 
-	  /* Skip intersected contours */
-	  if (curc->Flags.status == ISECTED)
-	    {
-	      prev = curc;
-	      continue;
-	    }
+      if (del_outside)
+        del_contour = curc->Flags.status != ISECTED &&
+                     !cntr_in_M_POLYAREA (curc, bpa, FALSE);
 
-	  /* Reset the intersection flags, since we keep these pieces */
-	  curc->Flags.status = UNKNWN;
+      /* Skip intersected contours */
+      if (curc->Flags.status == ISECTED) {
+        prev = curc;
+        continue;
+      }
 
-	  if (del_contour || hole_contour)
-	    {
+      /* Reset the intersection flags, since we keep these pieces */
+      curc->Flags.status = UNKNWN;
 
-	      remove_contour (a, prev, curc, !(is_first && is_last));
+      if (del_contour || hole_contour) {
 
-	      if (del_contour)
-		{
-		  /* Delete the contour */
-		  poly_DelContour (&curc);	/* NB: Sets curc to NULL */
-		}
-	      else if (hole_contour)
-		{
-		  /* Link into the list of holes */
-		  curc->next = *holes;
-		  *holes = curc;
-		}
-	      else
-		{
-		  assert (0);
-		}
+        remove_contour (a, prev, curc, !(is_first && is_last));
 
-	      if (is_first && is_last)
-		{
-		  remove_polyarea (pieces, a);
-		  poly_Free (&a);	/* NB: Sets a to NULL */
-		}
+        if (del_contour) {
+          /* Delete the contour */
+          poly_DelContour (&curc); /* NB: Sets curc to NULL */
+        } else if (hole_contour) {
+          /* Link into the list of holes */
+          curc->next = *holes;
+          *holes = curc;
+        } else {
+          assert (0);
+        }
 
-	    }
-	  else
-	    {
-	      /* Note the item we just didn't delete as the next
-	         candidate for having its "next" pointer adjusted.
-	         Saves walking the contour list when we delete one. */
-	      prev = curc;
-	    }
+        if (is_first && is_last) {
+          remove_polyarea (pieces, a);
+          poly_Free (&a); /* NB: Sets a to NULL */
+        }
 
-	  /* If we move or delete an outer contour, we need to move any holes
-	     we wish to keep within that contour to the holes list. */
-	  if (is_outline && del_contour)
-	    hole_contour = 1;
+      } else {
+        /* Note the item we just didn't delete as the next
+           candidate for having its "next" pointer adjusted.
+           Saves walking the contour list when we delete one. */
+        prev = curc;
+      }
 
-	}
+      /* If we move or delete an outer contour, we need to move any holes
+         we wish to keep within that contour to the holes list. */
+      if (is_outline && del_contour)
+        hole_contour = 1;
 
-      /* If we deleted all the pieces of the polyarea, *pieces is NULL */
     }
-  while ((a = anext), *pieces != NULL && !finished);
+
+  /* If we deleted all the pieces of the polyarea, *pieces is NULL */
+  } while ((a = anext), *pieces != NULL && !finished);
 }
 
 static void
 M_POLYAREA_Collect_separated (jmp_buf * e, PLINE * afst, POLYAREA ** contours,
-			      PLINE ** holes, int action, BOOLp maybe)
+                              PLINE ** holes, int action, BOOLp maybe)
 {
   PLINE **cur, **next;
 
-  for (cur = &afst; *cur != NULL; cur = next)
-    {
-      next = &((*cur)->next);
-      /* if we disappear a contour, don't advance twice */
-      if (cntr_Collect (e, cur, contours, holes, action, NULL, NULL, NULL))
-	next = cur;
-    }
+  for (cur = &afst; *cur != NULL; cur = next) {
+    next = &((*cur)->next);
+    /* if we disappear a contour, don't advance twice */
+    if (cntr_Collect (e, cur, contours, holes, action, NULL, NULL, NULL))
+      next = cur;
+  }
 }
 
 static void
@@ -2254,7 +2198,7 @@ M_POLYAREA_Collect (jmp_buf * e, POLYAREA * afst, POLYAREA ** contours,
 		    PLINE ** holes, int action, BOOLp maybe)
 {
   POLYAREA *a = afst;
-  POLYAREA *parent = NULL;	/* Quiet compiler warning */
+  POLYAREA *parent = NULL; /* Quiet compiler warning */
   PLINE **cur, **next, *parent_contour;
 
   assert (a != NULL);
@@ -2263,35 +2207,35 @@ M_POLYAREA_Collect (jmp_buf * e, POLYAREA * afst, POLYAREA ** contours,
   do
     {
       if (maybe && a->contours->Flags.status != ISECTED)
-	parent_contour = a->contours;
+          parent_contour = a->contours;
       else
-	parent_contour = NULL;
+          parent_contour = NULL;
 
       /* Take care of the first contour - so we know if we
        * can shortcut reparenting some of its children
        */
       cur = &a->contours;
       if (*cur != NULL)
-	{
-	  next = &((*cur)->next);
-	  /* if we disappear a contour, don't advance twice */
-	  if (cntr_Collect (e, cur, contours, holes, action, a, NULL, NULL))
-	    {
-	      parent = (*contours)->b;	/* InsCntr inserts behind the head */
-	      next = cur;
-	    }
-	  else
-	    parent = a;
-	  cur = next;
-	}
-      for (; *cur != NULL; cur = next)
-	{
-	  next = &((*cur)->next);
-	  /* if we disappear a contour, don't advance twice */
-	  if (cntr_Collect (e, cur, contours, holes, action, a, parent,
-			    parent_contour))
-	    next = cur;
-	}
+        {
+          next = &((*cur)->next);
+          /* if we disappear a contour, don't advance twice */
+          if (cntr_Collect (e, cur, contours, holes, action, a, NULL, NULL))
+            {
+              parent = (*contours)->b; /* InsCntr inserts behind the head */
+              next = cur;
+            }
+          else
+            parent = a;
+          cur = next;
+        }
+      for ( ; *cur != NULL; cur = next)
+        {
+          next = &((*cur)->next);
+          /* if we disappear a contour, don't advance twice */
+          if (cntr_Collect (e, cur, contours, holes, action, a, parent,
+                            parent_contour))
+            next = cur;
+        }
     }
   while ((a = a->f) != afst);
 }
@@ -2397,17 +2341,16 @@ poly_Boolean_free (POLYAREA * ai, POLYAREA * bi, POLYAREA ** res, int action)
       M_POLYAREA_update_primary (&e, res, &holes, action, b);
       M_POLYAREA_separate_isected (&e, res, &holes, &a_isected);
       M_POLYAREA_label_separated (a_isected, b, FALSE);
-      M_POLYAREA_Collect_separated (&e, a_isected, res, &holes, action,
-				    FALSE);
+      M_POLYAREA_Collect_separated (&e, a_isected, res, &holes, action, FALSE);
       M_B_AREA_Collect (&e, b, res, &holes, action);
       poly_Free (&b);
 
       /* free a_isected */
       while ((p = a_isected) != NULL)
-	{
-	  a_isected = p->next;
-	  poly_DelContour (&p);
-	}
+      {
+        a_isected = p->next;
+        poly_DelContour (&p);
+      }
 
       InsertHoles (&e, *res, &holes);
     }
@@ -2602,10 +2545,10 @@ poly_DelContour (PLINE ** c)
     {
       prev = cur->prev;
       if (cur->cvc_next != NULL)
-	{
-	  free (cur->cvc_next);
-	  free (cur->cvc_prev);
-	}
+        {
+          free (cur->cvc_next);
+          free (cur->cvc_prev);
+        }
       free (cur);
     }
   if ((*c)->head.cvc_next != NULL)
@@ -2803,7 +2746,7 @@ poly_Copy1 (POLYAREA * dst, const POLYAREA * src)
     {
       if (!poly_CopyContour (last, cur))
 	return FALSE;
-      r_insert_entry (dst->contour_tree, (BoxTypePtr) * last, 0);
+      r_insert_entry (dst->contour_tree, (BoxTypePtr) *last, 0);
       last = &(*last)->next;
     }
   return TRUE;
@@ -3002,7 +2945,7 @@ poly_Create (void)
 }
 
 void
-poly_FreeContours (PLINE ** pline)
+poly_FreeContours (PLINE **pline)
 {
   PLINE *pl;
